### DISCUSSION TITLE
tiling images and creating a labeled index image

### DISCUSSION THREAD

**Discussion by bruce-simonson** [ImageMagick/ImageMagick#2547] [Help] [answered]
Greetings,

I have been able to tile my image, using the code recommended here:  [split image into tiles](https://www.imagemagick.org/discourse-server/viewtopic.php?p=151589#p151589)

Basically, the code to tile an image, with borders around the tiles, and reconstituting the image is, for example:

> convert denali.tif -crop 20x20 -border 1x1 +adjoin denali_tile_%02d.gif
> convert denali_tile_*.gif -layers merge denali_tile_index.gif

I would like to label each of the tiles on the reconstituted index file, with the tile number.  Since the tile files are all saved in the above commands, perhaps using the tile file names would work?

I can force labels onto the final index file using annotate, with careful chosen offsets, but I'm wondering if there is a way to accomplish this without resorting to writing a separate program to do this annotation.  

It would be handy, and helpful, to include the desired annotation with options in the above code snippet, if possible.

Thanks for your help.

<img width="432" alt="tile_labeling_question" src="https://user-images.githubusercontent.com/66903636/92738629-f6093600-f330-11ea-92e9-8d8759871095.png">

**Comment by GeeMack** [comment, marked as answer]
This result can be achieved pretty simply with ImageMagick if you have the option to crop according to how many tiles you want instead of cropping to a certain size of tile.

This command would crop your input into a grid 10 tiles wide by 5 tiles high, shave 1 pixel off all the edges and add a 1 pixel border, write the 50 intermediate files as "denali_tile_00.gif",  "denali_tile_01.gif", etc., label all the tiles with sequential numbering, and re-assemble them to output "denali_tile_index.gif".

    convert denali.tif -crop 10x5@ -shave 1 -border 1 +adjoin -write denali_tile_%02d.gif ^
       -gravity center -fill white -pointsize 24 +repage -annotate +0+0 %[fx:t] +gravity ^
       +append -crop 5x1@ -append +repage denali_tile_index.gif

– Edited to Add –
In the first line there the "@" flag tells the crop operation to cut the image into 50 tiles, 10 tiles wide by 5 tiles high. Then it writes all those tiles to 50 individual files named "denali_tile_00.gif",  "denali_tile_01.gif", etc. before continuing with the command.

The second line sets the gravity to "center" so the annotations land in the middle of the tiles. It sets a fill color and pointsize for the annotation. You can put "-font YourFont" there, too, if you need a special font. Most importantly it un-sets the gravity with "+gravity", otherwise the following crops won't work correctly.

The third line appends all 50 of the images into a single horizontal row with "+append. Then it crops that row horizontally into 5 pieces wide by one high with "-crop 5x1@". The 5 pieces are assembled vertically with "-append" to produce the final image. At the end use "+repage" to reset the geometry of the output image to WxH+0+0.

If you want to make, for example, 4 tiles wide by 3 tiles high, you'd make that first crop "-crop 4x3@" and the second crop "-crop 3x1@".
– – – – – –

This is in Windows syntax. If you're on a *nix system, change those continued-line carets "^" to backslashes "\\".

If you don't need those individual tiles just omit the "-write denali_tile_%02d.gif ".

If the input image must be cropped into tiles of certain dimensions instead of a  particular number of tiles, the task becomes somewhat more complicated and we'd have to consider a different approach.

**Reply by bruce-simonson** [reply]
Excellent, I'm impressed. Thanks.

Two quick questions:

    What does the 2nd crop do (in the third line of code)?

    For clarity, can you add an example, that divides the image into 4 columns, and 3 rows, and creates the labeled index?

Thanks!

**Reply by GeeMack** [reply]
I added an explanation to my above comment that should answer those concerns.
